src/history/test/HistoryTests.cpp

```diff
@@ -320,6 +346,100 @@ TEST_CASE("History publish to multiple archives", "[history]")
     REQUIRE(catchupSimulation.catchupOffline(catchupApp, checkpointLedger));
 }
 
+TEST_CASE("Publish works correctly post shadow removal", "[history]")
+{
+    // Check levels (l, m] that must have resolved outputs
+    auto requireLevelsHaveOutput = [](uint32_t begin, uint32_t end,
+                                      HistoryArchiveState const& has) {
+        for (uint32_t i = 0; i < BucketList::kNumLevels; ++i)
+        {
+            auto next = has.currentBuckets[i].next;
+            if (i > begin && i <= end)
+            {
+                REQUIRE(next.hasOutputHash());
+            }
+            else
+            {
+                REQUIRE(next.isClear());
+            }
+        }
+    };
+
+    auto checkClearFutureBuckets =
+        [&](Application& app, uint32_t checkpoint,
+            uint32_t expectedLevelsCleared,
+            uint32_t maxLevelTouched /* levels touched by the upgrade */) {
+            REQUIRE(expectedLevelsCleared <= maxLevelTouched);
+
+            // First, ensure that latest HAS has futures cleared
+            auto w = app.getWorkScheduler()
+                         .executeWork<GetHistoryArchiveStateWork>();
+            auto has = w->getHistoryArchiveState();
+            REQUIRE(w->getState() == BasicWork::State::WORK_SUCCESS);
+            requireLevelsHaveOutput(expectedLevelsCleared, maxLevelTouched,
+                                    has);
+
+            // Second, pull previous [before upgrade] HAS, and
+            // make sure it contains resolved futures (as per < 12 protocols)
+            auto prevCheckpoint =
+                checkpoint - app.getHistoryManager().getCheckpointFrequency();
+            w = app.getWorkScheduler().executeWork<GetHistoryArchiveStateWork>(
+                prevCheckpoint);
+            auto oldHas = w->getHistoryArchiveState();
+            REQUIRE(w->getState() == BasicWork::State::WORK_SUCCESS);
+            requireLevelsHaveOutput(0, maxLevelTouched, oldHas);
+        };
+
+    auto configurator =
+        std::make_shared<RealGenesisTmpDirHistoryConfigurator>();
+    CatchupSimulation catchupSimulation{VirtualClock::VIRTUAL_TIME,
+                                        configurator};
+
+    uint32_t oldProto = Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED - 1;
+
+    SECTION("new-style merges begin in between checkpoints")
+    {
+        // Upgrade to oldProto
+        catchupSimulation.generateRandomLedger(oldProto);
+
+        uint32_t upgradeLedger = 64;
+        catchupSimulation.setProto12UpgradeLedger(upgradeLedger);
+        auto checkpointLedger = catchupSimulation.getLastCheckpointLedger(2);
+        catchupSimulation.ensureOfflineCatchupPossible(checkpointLedger);
+        auto levelsCleared = numLevelsSpills(upgradeLedger, checkpointLedger);
+        checkClearFutureBuckets(catchupSimulation.getApp(), checkpointLedger,
+                                levelsCleared, 3);
+    }
+    SECTION("new-style merges begin mid-checkpoint")
```

Comment:
Is the only difference between this section and the previous section `64` vs. `80`? If so, I'm not really sure how that corresponds to the semantic distinction between the phrases "in between checkpoints" and "mid-checkpoint". Those two phrases both seem to me like they mean the same thing, and if they don't, I don't know how they relate to `64` or `80`.

(And: if all 3 sections are identical except for the ledger number, maybe make it a loop?)

Reply:
fixed